### PR TITLE
Feature: JSON Artifacts

### PR DESCRIPTION
**Pull request by aaazzam**
Mostly a refactor of our public artifact methods.

The biggest change is a new public class-based interface to Artifacts, which:
- defaults the artifact type to `json`. 
- enables a `get` by key, which was previously available only through the CLI. 
- enables a `get_or_create` by key, which does what you think it does. 

I'm doing a draft PR 

- because there are _no tests written_, 
- to get feedback on the DevEx, and 
- there's _some_ UI work to be done so that it doesn't just say "Unknown" everywhere. 

Screenshot 2024-03-13 at 6 32 25 PM
e738d07-beed-4be4-99cc-c78abb709ae6">

### CODE REVIEW

File: src/prefect/artifacts.py [hunk 1]
```diff
@@ -2,21 +2,170 @@
...
+            )
+
+    async def format(self) -> Optional[Union[Dict[str, Any], Any]]:
+        return json.dumps(self.data)
+
+
+class LinkArtifact(Artifact):
+    link: str
+    link_text: Optional[str]
+    type: Optional[str] = "markdown"
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:128]:
```suggestion
    type: str = "link"
```

File: src/prefect/artifacts.py [hunk 2]
```diff
@@ -2,21 +2,170 @@
...
+        return (
+            f"[{self.link_text}]({self.link})"
+            if self.link_text
+            else f"[{self.link}]({self.link})"
+        )
+
+
+class MarkdownArtifact(Artifact):
+    markdown: str
+    type: Optional[str] = "markdown"
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:140]:
```suggestion
    type: str = "markdown"
```

File: src/prefect/artifacts.py [hunk 3]
```diff
@@ -2,21 +2,170 @@
...
+
+    async def format(self) -> str:
+        return self.markdown
+
+
+class TableArtifact(Artifact):
+    table: Union[Dict[str, List[Any]], List[Dict[str, Any]], List[List[Any]]]
+    type: Optional[str] = "table"
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:148]:
```suggestion
    type: str = "table"
```

File: src/prefect/artifacts.py [hunk 4]
```diff
@@ -87,15 +224,13 @@ async def create_link_artifact(
     Returns:
         The table artifact ID.
     """
-    formatted_link = f"[{link_text}]({link})" if link_text else f"[{link}]({link})"
-    artifact = await _create_artifact(
+    return await LinkArtifact(
+        type=type,
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:228]:
```suggestion
        type=LinkArtifact.type,
```

File: src/prefect/artifacts.py [hunk 5]
```diff
@@ -117,14 +252,12 @@ async def create_markdown_artifact(
     Returns:
         The table artifact ID.
     """
-    artifact = await _create_artifact(
+    return await MarkdownArtifact(
+        type=type,
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:256]:
```suggestion
        type=MarkdownArtifact.type,
```

**Review by jakekaplan — COMMENTED** [review]
Makes sense to me! I like using  a class method for creation. It makes it feel more like it's making a _thing_

File: src/prefect/artifacts.py [hunk 6]
```diff
@@ -2,21 +2,170 @@
...
+                    sort=ArtifactSort.UPDATED_DESC,
+                    artifact_filter=ArtifactFilter(key=ArtifactFilterKey(any_=[key])),
+                )
+            ),
+            None,
+        )
+
+    @classmethod
+    @sync_compatible
+    async def get_or_create(
```
[the remark this answers is not in the document]
Comment (abrookins) [reply src/prefect/artifacts.py:90]: Hmmm -- does that interface make sense at all, then? If the data model is such that multiple artifacts would exist, it should be, like, `get_latest_or_create` or `get_or_create(return_latest=True)`? 🤔

File: src/prefect/artifacts.py [hunk 7]
```diff
@@ -2,21 +2,173 @@
...
+    An artifact is a piece of data that is created by a flow or task run.
+    https://docs.prefect.io/latest/concepts/artifacts/
+
+    Arguments:
+        type: A string identifying the type of artifact.
+        key: A user-provided string identifier.
+          The key must only contain lowercase letters, numbers, and dashes.
+        description: A user-specified description of the artifact.
+        data: A JSON payload that allows for a result to be retrieved.
+    """
```
Comment (serinamarie) [inline src/prefect/artifacts.py:35]: should we note the optional args here?

File: src/prefect/artifacts.py [hunk 8]
```diff
@@ -2,21 +2,173 @@
...
+    @classmethod
+    @sync_compatible
+    async def get_or_create(
+        cls,
+        key: Optional[str] = None,
+        description: Optional[str] = None,
+        data: Optional[Union[Dict[str, Any], Any]] = None,
+        client: Optional[PrefectClient] = None,
+        **kwargs: Any,
+    ) -> Tuple[ArtifactResponse, bool]:
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:100]: What would think about this just returning the artifact by default?

File: src/prefect/artifacts.py [hunk 9]
```diff
@@ -2,21 +2,173 @@
...
+        Arguments:
+            key (str, optional): The key of the artifact to get.
+            client (PrefectClient, optional): The PrefectClient
+
+        Returns:
+            (ArtifactResponse, optional): The artifact (if found).
+        """
+        client, _ = get_or_create_client(client)
+        return next(
+            iter(
```
Comment (jakekaplan) [inline src/prefect/artifacts.py:81]: fancy

**Pull request merged**